test/serve/build_delay_test.dart

```diff
@@ -72,52 +54,107 @@ main() {
+
+    // Regression test for https://github.com/dart-lang/sdk/issues/29890
+    test("editors safe write features shouldn't cause failed builds", () async {
+      // Simulate the safe-write feature from many editors:
+      //   - Create a backup file
+      //   - Edit original file
+      //   - Delete backup file
+      var backupFilePath = p.join(d.sandbox, appPath, "lib", "lib.dart.bak");
+      writeTextFile(backupFilePath, "foo() => 'foo';");
+      libWatcher.addEvent(new WatchEvent(ChangeType.ADD, backupFilePath));
+      await new Future(() {});
+      deleteEntry(backupFilePath);
+      libWatcher.addEvent(new WatchEvent(ChangeType.REMOVE, backupFilePath));
+
+      // Should get a single successful build result.
+      await _waitForBarback();
+      expect(barbackResults.length, 1);
+      expect(barbackResults.first.succeeded, isTrue);
+    });
+  });
+}
 
-    await requestShouldSucceed("packages/myapp/lib.dart", "foo() => '9';");
+/// Mock [WatcherType] that creates [_MockDirectoryWatcher]s and gives you
+/// access to them.
+class _MockWatcherType implements WatcherType {
+  final watchers = <String, _MockDirectoryWatcher>{};
 
-    await endPubServe();
-  });
+  _MockDirectoryWatcher create(String dir) {
+    var watcher = new _MockDirectoryWatcher(dir);
+    watchers[dir] = watcher;
+    return watcher;
+  }
+}
 
-  // Regression test for https://github.com/dart-lang/sdk/issues/29890
-  test("editors safe write features shouldn't cause failed builds", () async {
-    var pubServeProcess = await pubServe(forcePoll: false);
-    expect(pubServeProcess.stdout,
-        emitsThrough(contains('Build completed successfully')));
-    await requestShouldSucceed("packages/myapp/lib.dart", "foo() => 'foo';");
+/// Mock [DirectoryWatcher] that you add events to manually using [addEvent].
+class _MockDirectoryWatcher implements DirectoryWatcher {
+  final _eventsController = new StreamController<WatchEvent>();
+  Stream<WatchEvent> get events => _eventsController.stream;
 
-    // Simulate the safe-write feature from many editors:
-    //   - Create a backup file
-    //   - Edit original file
-    //   - Delete backup file
-    var backupFilePath = p.join(d.sandbox, appPath, "lib", "lib.dart.bak");
-    writeTextFile(backupFilePath, "foo() => 'foo';");
-    // Allow pub to schedule a new build.
-    await new Future(() {});
-    writeTextFile(libFilePath, "foo() => 'bar';");
-    deleteEntry(backupFilePath);
+  final String path;
+  String get directory => path;
 
-    expect(pubServeProcess.stderr, neverEmits(contains("Build error")));
-    expect(pubServeProcess.stdout,
-        emitsThrough(contains('Build completed successfully')));
-    await requestShould404("packages/myapp/lib.dart.bak");
-    await requestShouldSucceed("packages/myapp/lib.dart", "foo() => 'bar';");
+  final _readyCompleter = new Completer<Null>()..complete();
+  Future<Null> get ready => _readyCompleter.future;
+  bool get isReady => _readyCompleter.isCompleted;
```

Comment:
done, and deleted _readyCompleter in favor of just a regular `Future` for `ready`.